DNM Playing with orchestrator

pyricau opened the pull request:
Just opening this to keep track of WIP.

julioz commented:
@pyricau Can you provide some more information on what's the hypothesis here and the plan you have to move this forward?
The issue with the orchestrator integration (#1046) prevents us from running tests with leak canary and we could allocate some time to contribute here if we can understand what exactly is the plan.

pyricau commented:
This is very much work in progress, but the key idea is to explore the code of Orchestrator and figure out how we can send failures for a successful test, from within a test run listener. Which essentially means using reflection and highjacking the way the normal test framework sends test results to the result gathering process. One key issue is figuring out what happens if two results are sent, and if that's an issue how to make sure only one is sent (one way being to be first to send the result then exiting the process)

pyricau commented:
More prototyping.

Basically I figured that we can replace the oboCallback with reflection. Unfortunately test listeners are called each on different threads so we need to hold the orchestrator listener thread, and later release it and allow it or not to forward the finished call.

One key issue at the moment: it looks like we're getting our "test finished" callback before the activity is destroyed. Might be a timing thing, ie activity destroy happens on main thread? Need to check if cleanup of activity rule blocks for activity destroy from instrumentation thread or just moves forward.

pyricau commented:
I added a call to activity.finish + sleep from the test, and moved forward past this (in the future  we can block wait for all activities to finish from the run listener)

However now I get:

Test running failed: Test run failed to complete. Expected 1 tests, received 0

When I pull the test report I do see one failure and it's the correct failure now. So, progress, but not quite there yet.

pyricau commented:
Closing, will open up clean PR
